File: app/src/main/java/com/vpulse/ftpnext/database/FTPServerTable/FTPServerDAO.java
package com.vpulse.ftpnext.database.FTPServerTable;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.vpulse.ftpnext.core.FTPCharacterEncoding;
import com.vpulse.ftpnext.core.FTPType;
import com.vpulse.ftpnext.core.LogManager;
import com.vpulse.ftpnext.database.ADataAccessObject;
import com.vpulse.ftpnext.ftpservices.FTPServices;

import java.util.List;

public class FTPServerDAO extends ADataAccessObject<FTPServer> implements IFTPServerSchema {

    private static final String TAG = "DATABASE : FTP Host DAO";

    private static final int VERSION_UPDATE_ABSOLUTE_PATH_VALUE = 3;

    public FTPServerDAO(SQLiteDatabase iDataBase) {
        super(iDataBase);
        LogManager.info(TAG, "Creating " + this.getClass().getSimpleName());
    }

    // TODO test
    public FTPServer fetchByName(String iName) {
        final String[] lSelectionArgs = {iName};
        final String lSelection = COLUMN_NAME + " = ?";
        FTPServer lObject = null;

        mCursor = super.query(TABLE, COLUMN_ARRAY, lSelection, lSelectionArgs, COLUMN_NAME);
        if (mCursor != null) {
            mCursor.moveToFirst();
            while (!mCursor.isAfterLast()) {
                lObject = cursorToEntity(mCursor);
                mCursor.moveToNext();
            }
            mCursor.close();
        }
        return lObject;
    }

    @Override
    public FTPServer fetchById(int iId) {
        return super.fetchById(TABLE, iId, COLUMN_DATABASE_ID, COLUMN_ARRAY);
    }

    @Override
    public List<FTPServer> fetchAll() {
        return super.fetchAll(TABLE, COLUMN_ARRAY, COLUMN_DATABASE_ID);
    }

    @Override
    public int add(FTPServer iObject) {
        if (iObject == null) {
            LogManager.error(TAG, "Object to add is null");
            return -1;
        }

        return super.add(iObject, TABLE);
    }

    @Override
    public boolean update(FTPServer iObject) {
        if (iObject == null)
            return LogManager.error(TAG, "Object to update is null");

        return super.update(iObject, iObject.getDataBaseId(), TABLE, COLUMN_DATABASE_ID);
    }

    @Override
    public boolean deleteAll() {
        return super.delete(TABLE, null, null) > 0;
    }

    @Override
    public boolean delete(int iId) {
        return super.delete(iId, TABLE, COLUMN_DATABASE_ID);
    }

    @Override
    public boolean delete(FTPServer iObject) {
        if (iObject == null) {
            LogManager.error(TAG, "Object to set content value is null");
            return false;
        }
        return delete(iObject.getDataBaseId());
    }

    @Override
    public void onUpgradeTable(int iOldVersion, int iNewVersion) {
        if (iOldVersion < VERSION_UPDATE_ABSOLUTE_PATH_VALUE) {
            List<FTPServer> lFTPServerList = fetchAll();

            for (FTPServer lItem : lFTPServerList) {
                if (!lItem.getAbsolutePath().endsWith("/")) {
                    lItem.setAbsolutePath(lItem.getAbsolutePath() + "/");
                    update(lItem);
                }
            }
        }
    }

    @Override
    protected void setContentValue(FTPServer iObject) {
        if (iObject == null) {
            LogManager.error(TAG, "Object to set content value is null");
            return;
        }

        mContentValues = new ContentValues();
        if (iObject.getDataBaseId() != 0) {
            mContentValues.put(COLUMN_DATABASE_ID, iObject.getDataBaseId());
        }
        mContentValues.put(COLUMN_NAME, iObject.getName());
        mContentValues.put(COLUMN_USER, iObject.getUser());
        mContentValues.put(COLUMN_PASS, iObject.getPass());
        mContentValues.put(COLUMN_SERVER, iObject.getServer());
        mContentValues.put(COLUMN_PORT, iObject.getPort());
        mContentValues.put(COLUMN_FOLDER_NAME, iObject.getFolderName());
        mContentValues.put(COLUMN_ABSOLUTE_PATH, iObject.getAbsolutePath());
        mContentValues.put(COLUMN_CHARACTER_ENCODING, iObject.getFTPCharacterEncoding().getValue());
        mContentValues.put(COLUMN_TYPE, iObject.getFTPType().getValue());
    }

    @Override
    protected FTPServer cursorToEntity(Cursor iCursor) {
        if (iCursor == null) {
            LogManager.error(TAG, "Cursor in cursorToEntity is null");
            return null;
        }

        FTPServer oObject = new FTPServer();
        if (iCursor.getColumnIndex(COLUMN_DATABASE_ID) != -1)
            oObject.setDataBaseId(iCursor.getInt(iCursor.getColumnIndexOrThrow(COLUMN_DATABASE_ID)));
        if (iCursor.getColumnIndex(COLUMN_NAME) != -1)
            oObject.setName(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_NAME)));
        if (iCursor.getColumnIndex(COLUMN_USER) != -1)
            oObject.setUser(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_USER)));
        if (iCursor.getColumnIndex(COLUMN_PASS) != -1)
            oObject.setPass(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_PASS)));
        if (iCursor.getColumnIndex(COLUMN_SERVER) != -1)
            oObject.setServer(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_SERVER)));
        if (iCursor.getColumnIndex(COLUMN_PORT) != -1)
            oObject.setPort(iCursor.getInt(iCursor.getColumnIndexOrThrow(COLUMN_PORT)));
        if (iCursor.getColumnIndex(COLUMN_FOLDER_NAME) != 1)
            oObject.setFolderName(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_FOLDER_NAME)));
        if (iCursor.getColumnIndex(COLUMN_ABSOLUTE_PATH) != -1)
            oObject.setAbsolutePath(iCursor.getString(iCursor.getColumnIndexOrThrow(COLUMN_ABSOLUTE_PATH)));
        if (iCursor.getColumnIndex(COLUMN_CHARACTER_ENCODING) != -1)
            oObject.setFTPCharacterEncoding(FTPCharacterEncoding.getValue(
                    iCursor.getInt(iCursor.getColumnIndexOrThrow(COLUMN_CHARACTER_ENCODING))));
        if (iCursor.getColumnIndex(COLUMN_TYPE) != -1)
            oObject.setFTPType(FTPType.getValue(
                    iCursor.getInt(iCursor.getColumnIndexOrThrow(COLUMN_TYPE))));

        return oObject;
    }
}

File: app/src/main/java/com/vpulse/ftpnext/database/FTPServerTable/FTPServerDAO.kt
package com.vpulse.ftpnext.database.FTPServerTable

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.vpulse.ftpnext.core.FTPCharacterEncoding
import com.vpulse.ftpnext.core.FTPType
import com.vpulse.ftpnext.core.LogManager
import com.vpulse.ftpnext.database.ADataAccessObject

class FTPServerDAO constructor(iDataBase: SQLiteDatabase) : ADataAccessObject<FTPServer>(iDataBase),
    IFTPServerSchema {

    init {
        LogManager.info(TAG, "Creating " + this.javaClass.simpleName)
    }

    // TODO test
    fun fetchByName(iName: String?): FTPServer {
        val lSelectionArgs: Array<String?> = arrayOf(iName)
        val lSelection: String = IFTPServerSchema.COLUMN_NAME + " = ?"
        var lObject: FTPServer? = null
        mCursor = super.query(
            IFTPServerSchema.TABLE,
            IFTPServerSchema.COLUMN_ARRAY,
            lSelection,
            lSelectionArgs,
            IFTPServerSchema.COLUMN_NAME
        )
        if (mCursor != null) {
            mCursor!!.moveToFirst()
            while (!mCursor!!.isAfterLast) {
                lObject = cursorToEntity(mCursor)
                mCursor!!.moveToNext()
            }
            mCursor!!.close()
        }
        return lObject
    }

    override fun fetchById(iId: Int): FTPServer? {
        return super.fetchById(
            IFTPServerSchema.TABLE,
            iId,
            IFTPServerSchema.COLUMN_DATABASE_ID,
            IFTPServerSchema.COLUMN_ARRAY
        )
    }

    override fun fetchAll(): List<FTPServer> {
        return super.fetchAll(
            IFTPServerSchema.TABLE,
            IFTPServerSchema.COLUMN_ARRAY,
            IFTPServerSchema.COLUMN_DATABASE_ID
        )
    }

    override fun add(iObject: FTPServer): Int {
        return super.add(iObject, IFTPServerSchema.Companion.TABLE)
    }

    override fun update(iObject: FTPServer): Boolean {
        return super.update(
            iObject, iObject.dataBaseId, IFTPServerSchema.TABLE, IFTPServerSchema.COLUMN_DATABASE_ID
        )
    }

    override fun deleteAll(): Boolean {
        return super.delete(IFTPServerSchema.Companion.TABLE, null, null) > 0
    }

    override fun delete(iId: Int): Boolean {
        return super.delete(
            iId, IFTPServerSchema.TABLE, IFTPServerSchema.COLUMN_DATABASE_ID
        )
    }

    override fun delete(iObject: FTPServer): Boolean {
        return delete(iObject.dataBaseId)
    }

    override fun onUpgradeTable(iOldVersion: Int, iNewVersion: Int) {
        if (iOldVersion < VERSION_UPDATE_ABSOLUTE_PATH_VALUE) {
            val lFTPServerList: List<FTPServer?> = fetchAll()
            for (lItem: FTPServer? in lFTPServerList) {
                if (!lItem!!.absolutePath!!.endsWith("/")) {
                    lItem.absolutePath = lItem.absolutePath + "/"
                    update(lItem)
                }
            }
        }
    }

    override fun setContentValue(iObject: FTPServer) {
        mContentValues = ContentValues()
        if (iObject.dataBaseId != 0) {
            mContentValues!!.put(
                IFTPServerSchema.Companion.COLUMN_DATABASE_ID, iObject.dataBaseId
            )
        }
        mContentValues!!.put(IFTPServerSchema.COLUMN_NAME, iObject.name)
        mContentValues!!.put(IFTPServerSchema.COLUMN_USER, iObject.user)
        mContentValues!!.put(IFTPServerSchema.COLUMN_PASS, iObject.pass)
        mContentValues!!.put(IFTPServerSchema.COLUMN_SERVER, iObject.server)
        mContentValues!!.put(IFTPServerSchema.COLUMN_PORT, iObject.port)
        mContentValues!!.put(IFTPServerSchema.COLUMN_FOLDER_NAME, iObject.folderName)
        mContentValues!!.put(
            IFTPServerSchema.COLUMN_ABSOLUTE_PATH, iObject.absolutePath
        )
        mContentValues!!.put(
            IFTPServerSchema.COLUMN_CHARACTER_ENCODING, iObject.fTPCharacterEncoding.value
        )
        mContentValues!!.put(
            IFTPServerSchema.COLUMN_TYPE, iObject.fTPType.value
        )
    }

    override fun cursorToEntity(iCursor: Cursor): FTPServer {
        val oObject = FTPServer()
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_DATABASE_ID) != -1) {
            oObject.dataBaseId = iCursor.getInt(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_DATABASE_ID)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_NAME) != -1) {
            oObject.name = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_NAME)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_USER) != -1) {
            oObject.user = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_USER)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_PASS) != -1) {
            oObject.pass = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_PASS)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_SERVER) != -1) {
            oObject.server = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_SERVER)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_PORT) != -1) {
            oObject.port = iCursor.getInt(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_PORT)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_FOLDER_NAME) != 1) {
            oObject.folderName = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_FOLDER_NAME)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_ABSOLUTE_PATH) != -1) {
            oObject.absolutePath = iCursor.getString(
                iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_ABSOLUTE_PATH)
            )
        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_CHARACTER_ENCODING) != -1) {
            oObject.fTPCharacterEncoding = FTPCharacterEncoding.getValue(
                iCursor.getInt(iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_CHARACTER_ENCODING))
            )

        }
        if (iCursor.getColumnIndex(IFTPServerSchema.COLUMN_TYPE) != -1) {
            oObject.fTPType = FTPType.getValue(
                iCursor.getInt(iCursor.getColumnIndexOrThrow(IFTPServerSchema.COLUMN_TYPE))
            )

        }
        return oObject
    }

    companion object {
        private const val TAG: String = "DATABASE : FTP Host DAO"
        private const val VERSION_UPDATE_ABSOLUTE_PATH_VALUE: Int = 3
    }
}